File: pascal/hanoi.pas
unit hanoi;

interface

uses
    hanoiRecord,
    graficos,
    stack;

procedure newGame(difficulty: integer);
procedure gameLoop(game: PHanoi);

implementation

const
    TOWER_A = 0;
    TOWER_B = 1;
    TOWER_C = 2;

    AB = 12;
    AC = 13;

    BA = 21;
    BC = 23;

    CA = 31;
    CB = 32;

    QUIT = 0;

procedure newGame(difficulty: integer);

var
    game: THanoi;
    i: integer;

begin
    game.size := difficulty;
    
    for i := 0 to 2 do
        game.towers[i] := stack.new();

    for i := difficulty downto 1 do
        stack.push(@game.towers[TOWER_A], i);

    gameLoop(@game);
end;

procedure gameLoop(game: PHanoi);

var
    gameRunning: boolean = true;
    temp, input: integer;
    srcTower, destTower: integer;
    src, dest: integer; 
    i: integer;
    auxNode, auxNext: ^TNode;
    auxStack: TStack;

begin
    printGame(game);

    while gameRunning do
    begin
        // Input and logic
        read(input);
        case input of
            AB, AC, BA, BC, CA, CB:
            begin
                srcTower := (input div 10) - 1;
                destTower := (input mod 10) - 1;

                src := stack.peek(@game^.towers[srcTower]);
                dest := stack.peek(@game^.towers[destTower]);

                if (src <> 0) and ((src < dest) or (dest = 0)) then
                begin
                    temp := stack.pop(@game^.towers[srcTower]);
                    stack.push(@game^.towers[destTower], temp);
                end;
            end;

            QUIT:
                gameRunning := false;
        end;
        
        // ----- Graphics -----
        printGame(game);

        // ----- Victory? -----
        if game^.towers[TOWER_C].size = game^.size then
        begin
            write(#10'Congratulations! Type any number to return to the main menu. ');
            read(i);
            gameRunning := false;
        end;
    end;

    // Free stack nodes
    for i := 0 to 2 do
    begin
        auxStack := game^.towers[i];
        auxNode := auxStack.top;

        while auxNode <> nil do
        begin
            auxNext := auxNode^.next;
            freeMem(auxNode);
            auxNode := auxNext;
        end;
    end;
end;

end.

File: rust/src/hanoi.rs
use crate::myprint;
use crate::stack::Stack;
use crate::util::read_i32;
use crate::graphics::print_game;

pub struct Hanoi {
    pub size: i32,
    pub towers: [Stack; 3]
}

const TOWER_A: usize = 0;
// const TOWER_B: usize = 1;
const TOWER_C: usize = 2;

const AB: i32 = 12;
const AC: i32 = 13;
const BA: i32 = 21;
const BC: i32 = 23;
const CA: i32 = 31;
const CB: i32 = 32;

const QUIT: i32 = 0;

pub fn new_game(difficulty: i32) {
    let mut game = Hanoi {
        size: difficulty,
        towers: [
            Stack::new(),
            Stack::new(),
            Stack::new()
        ]
    };

    for i in (1..=difficulty).rev() {
        game.towers[TOWER_A].push(i);
    }

    game_loop(game);
}

fn game_loop(mut game: Hanoi) {
    let mut game_running: bool = true;
    print_game(&mut game);
    //read_i32();
    while game_running {
        let input = read_i32();
        match input {
            AB | AC | BA | BC | CA | CB => {
                let src_tower = ((input / 10) -1) as usize;
                let dest_tower = ((input % 10) - 1) as usize;

                let src = game.towers[src_tower].peek();
                let dest = game.towers[dest_tower].peek();

                if (src != 0) && ((src < dest) || (dest == 0)) {
                    let temp = game.towers[src_tower].pop();
                    game.towers[dest_tower].push(temp);
                }
            }

            QUIT => {
                game_running = false;
            }

            _ => {}
        }

        // ----- Graphics -----
        print_game(&mut game);

        if game.towers[TOWER_C].size == game.size {
            myprint!("\nCongratulations! Insert any number to start a new game. ");
            read_i32();
            game_running = false;
        }
    }

}